pkg/kubectl/cmd/create_role.go

```diff
@@ -0,0 +1,255 @@
+		}
+
+		c.Resources = append(c.Resources, schema.GroupVersionResource{Group: apiGroup, Resource: resourceName})
+	}
+
+	return nil
+}
+
+func (c *CreateRoleOptions) Validate(f cmdutil.Factory) error {
+	if c.Name == "" {
+		return fmt.Errorf("name must be specified")
+	}
+
+	// validate verbs.
+	if len(c.Verbs) == 0 {
+		return fmt.Errorf("at least one verb must be specified")
+	}
+
+	for _, v := range c.Verbs {
+		if !arrayContains(validResourceVerbs, v) && !arrayContains(validNonResourceVerbs, v) {
+			return fmt.Errorf("invalid verb: '%s'", v)
+		}
+	}
+
+	// validate resources.
+	mapper, _ := f.Object()
+
+	if len(c.Resources) == 0 {
+		return fmt.Errorf("at least one resource must be specified")
+	}
+
+	for _, r := range c.Resources {
+		_, err := mapper.ResourceFor(r)
+		if err != nil {
+			return err
+		}
+	}
+
+	// validate resource names, can not apply resource names to multiple resources.
+	if len(c.ResourceNames) > 0 && len(c.Resources) > 1 {
+		return fmt.Errorf("resource name(s) can not be applied to multiple resources")
+	}
+
+	return nil
+}
+
+func (c *CreateRoleOptions) RunCreateRole(f cmdutil.Factory, cmdOut io.Writer, cmd *cobra.Command, args []string) error {
+	mapper, typer := f.Object()
+	dryRun, outputFormat := cmdutil.GetDryRunFlag(cmd), cmdutil.GetFlagString(cmd, "output")
+
+	// Remove duplicate verbs.
+	verbs, containsVerbAll := []string{}, false
+	for _, v := range c.Verbs {
+		if !arrayContains(verbs, v) {
+			verbs = append(verbs, v)
+		}
+		if v == "*" {
+			containsVerbAll = true
+		}
+	}
+	// VerbAll respresents all kinds of verbs.
+	if containsVerbAll {
+		verbs = []string{"*"}
+	}
+
+	// Remove duplicate resources and transfer resource name from short form to long form.
+	groupResourceMapping := map[string][]string{}
+	for _, r := range c.Resources {
+		resource, err := mapper.ResourceFor(r)
+		if err != nil {
+			return err
+		}
+		if !arrayContains(groupResourceMapping[resource.Group], resource.Resource) {
+			groupResourceMapping[resource.Group] = append(groupResourceMapping[resource.Group], resource.Resource)
+		}
+	}
+
+	// Remove duplicate resource names.
+	resourceNames := []string{}
+	for _, n := range c.ResourceNames {
+		if !arrayContains(resourceNames, n) {
+			resourceNames = append(resourceNames, n)
+		}
+	}
+
+	role := &rbac.Role{}
+
+	// Create separate rule for each of the api group.
+	rules := []rbac.PolicyRule{}
+	for g, resources := range groupResourceMapping {
+		rule := rbac.PolicyRule{}
+		rule.Verbs = verbs
+		rule.Resources = resources
+		rule.APIGroups = []string{g}
+		rule.ResourceNames = resourceNames
+		rules = append(rules, rule)
+	}
+	role.Name = c.Name
+	role.Rules = rules
+
+	// Create role.
+	namespace, _, err := f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+
+	gvks, _, err := typer.ObjectKinds(role)
+	if err != nil {
+		return err
+	}
+	gvk := gvks[0]
+	mapping, err := mapper.RESTMapping(schema.GroupKind{Group: gvk.Group, Kind: gvk.Kind}, gvk.Version)
```

Comment:
You know exactly what you're going to create.  Just call `f.ClientSet` and select the one you want.